TLM/TLM/U/BaseUButton.cs

```diff
@@ -0,0 +1,109 @@
+﻿namespace TrafficManager.U {
+    using System;
+    using ColossalFramework.UI;
+    using CSUtil.Commons;
+    using TrafficManager.State.Keybinds;
+    using TrafficManager.Util;
+    using UnityEngine;
+
+    /// <summary>
+    /// A smart button which can change its foreground and background textures based on its state,
+    /// also has a localised tooltip.
+    /// This is an abstract base class for buttons.
+    /// </summary>
+    public abstract class BaseUButton : UIButton {
+        public abstract bool CanActivate();
+
+        public abstract string ButtonName { get; }
+
+        public abstract string FunctionName { get; }
+
+        public abstract string[] FunctionNames { get; }
+
+        public abstract Texture2D AtlasTexture { get; }
+
+        public abstract int Width { get; }
+
+        public abstract int Height { get; }
+
+        public override void Start() {
+            atlas = U.ButtonTexture.CreateAtlas(
+                ButtonName,
+                FunctionNames,
+                CanActivate(),
+                AtlasTexture);
+            m_ForegroundSpriteMode = UIForegroundSpriteMode.Scale;
+            UpdateProperties();
+
+            // Enable button sounds.
+            playAudioEvents = true;
+        }
+
+        public abstract bool Active { get; }
```

Comment:
Note: My road selection also provides an optional disabled state:
https://github.com/kianzarrin/Cities-Skylines-Traffic-Manager-President-Edition/blob/cb19aa340e2f6e22e24f1bca015a0191ea84ba80/TLM/TLM/UI/LinearSpriteButton.cs#L27

Reply:
Yeah this entire thing is convoluted, i want a more generic button with all these states, now also including the disabled state. Will have remember to take your change in too, whether your or mine gets merged first (your possibly goes in first).